cli/mender-artifact/write.go

```diff
@@ -460,3 +471,125 @@ func extractKeyValues(params []string) (*map[string]string, error) {
 	}
 	return keyValues, nil
 }
+
+// SSH to remote host and dump rootfs snapshot to a local temporary file.
+func getDeviceSnapshot(host string) (string, error) {
+	var userAtHost string
+	port := "22"
+	if remotePort := strings.Split(host, ":"); len(remotePort) == 2 {
+		port = remotePort[1]
+		userAtHost = remotePort[0]
+	} else {
+		userAtHost = host
+	}
+
+	// First echo to stdout such that we know when ssh connection is
+	// established (password prompt is written to /dev/tty directly,
+	// and hence impossible to detect).
+	args := []string{"-p", port, userAtHost, "sudo", "-S", "/bin/sh", "-c",
+		"'echo \"aLL SET!\"; mender snapshot dump | cat'"}
+
+	cmd := exec.Command("ssh", args...)
+
+	// Simply connect stdin/stderr
+	cmd.Stdin = os.Stdin
+	cmd.Stderr = os.Stderr
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return "", errors.New("Error redirecting stdout on exec")
+	}
+
+	if err := cmd.Start(); err != nil {
+		return "", err
+	}
+
+	// Wait for 60 seconds for ssh to establish connection
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancel()
+	errChan := make(chan error)
+	go func() {
+		stdoutRdr := bufio.NewReader(stdout)
+		for {
+			line, err := stdoutRdr.ReadString('\n')
+			if err != nil {
+				errChan <- errors.Wrap(err, "Error waiting for "+
+					"device to start dumping snapshot")
+				return
+			}
+			if strings.Contains(line, "aLL SET!") {
+				break
+			}
+		}
+		errChan <- nil
+	}()
+	select {
+	case err = <-errChan:
+		// Go routine finished
+
+	case <-ctx.Done():
+		// Context deadline exceeded
+		err = ctx.Err()
+	}
+
+	if err != nil {
+		cmd.Process.Kill()
+		return "", err
+	}
+	f, err := ioutil.TempFile("", "rootfs.ext4")
```

Comment:
Btw, we need some kind of cleanup handling for this. I got failures for various reasons while playing with it, and these files then remain. They are often multi gigabyte files, so they should be cleaned up if there is a failure.